File: hushh-webapp/ios/App/App/Plugins/HushhSyncPlugin.swift
import Capacitor

/**
 * HushhSyncPlugin - Cloud Synchronization (Capacitor 8)
 * Port of Android HushhSyncPlugin.kt
 */
@objc(HushhSyncPlugin)
public class HushhSyncPlugin: CAPPlugin, CAPBridgedPlugin {
    
    // MARK: - CAPBridgedPlugin Protocol
    public let identifier = "HushhSyncPlugin"
    public let jsName = "HushhSync"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "sync", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "push", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pull", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "syncVault", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSyncStatus", returnType: CAPPluginReturnPromise)
    ]
    
    private let TAG = "HushhSync"
    private let defaultBackendUrl = "https://consent-protocol-1006304528804.us-central1.run.app"
    
    private lazy var urlSession: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 30
        return URLSession(configuration: config)
    }()
    
    // MARK: - Sync
    @objc func sync(_ call: CAPPluginCall) {
        let authToken = call.getString("authToken")
        
        // Placeholder: push + pull
        let pushResult = performPush(authToken: authToken)
        let pullResult = performPull(authToken: authToken)
        
        call.resolve([
            "success": true,
            "pushedRecords": pushResult,
            "pulledRecords": pullResult,
            "conflicts": 0,
            "timestamp": Int64(Date().timeIntervalSince1970 * 1000)
        ])
    }
    
    // MARK: - Push
    @objc func push(_ call: CAPPluginCall) {
        let authToken = call.getString("authToken")
        let pushedRecords = performPush(authToken: authToken)
        
        call.resolve([
            "success": true,
            "pushedRecords": pushedRecords
        ])
    }
    
    // MARK: - Pull
    @objc func pull(_ call: CAPPluginCall) {
        let authToken = call.getString("authToken")
        let pulledRecords = performPull(authToken: authToken)
        
        call.resolve([
            "success": true,
            "pulledRecords": pulledRecords
        ])
    }
    
    // MARK: - Sync Vault
    @objc func syncVault(_ call: CAPPluginCall) {
        guard let userId = call.getString("userId") else {
            call.reject("Missing required parameter: userId")
            return
        }
        
        let authToken = call.getString("authToken")
        let backendUrl = call.getString("backendUrl") ?? defaultBackendUrl
        
        guard let url = URL(string: "\(backendUrl)/api/sync/vault") else {
            call.reject("Invalid URL")
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = authToken {
            request.addValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: ["userId": userId])
        } catch {
            call.reject("Failed to encode body")
            return
        }
        
        urlSession.dataTask(with: request) { _, response, error in
            if let httpResponse = response as? HTTPURLResponse {
                call.resolve(["success": (200...299).contains(httpResponse.statusCode)])
            } else {
                call.resolve(["success": false])
            }
        }.resume()
    }
    
    // MARK: - Get Sync Status
    @objc func getSyncStatus(_ call: CAPPluginCall) {
        // Placeholder - no local SQLCipher yet
        call.resolve([
            "pendingCount": 0,
            "lastSyncTimestamp": 0,
            "hasPendingChanges": false
        ])
    }
    
    // MARK: - Private Helpers
    private func performPush(authToken: String?) -> Int {
        print("🔄 [\(TAG)] Push completed")
        return 0
    }
    
    private func performPull(authToken: String?) -> Int {
        print("🔄 [\(TAG)] Pull completed")
        return 0
    }
}

File: hushh-webapp/android/app/src/main/java/com/hushh/app/plugins/HushhSync/HushhSyncPlugin.kt
package com.hushh.app.plugins.HushhSync

import android.util.Log
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

/**
 * Hushh Sync Plugin - Cloud Synchronization
 * Port of iOS HushhSyncPlugin.swift
 *
 * Handles local-cloud data synchronization
 */
@CapacitorPlugin(name = "HushhSync")
class HushhSyncPlugin : Plugin() {

    private val TAG = "HushhSync"
    private val httpClient = OkHttpClient()

    // Cloud Run backend URL
    private val backendUrl: String
        get() = System.getenv("BACKEND_URL") ?: "https://consent-protocol-1006304528804.us-central1.run.app"

    // ==================== Sync ====================

    @PluginMethod
    fun sync(call: PluginCall) {
        val authToken = call.getString("authToken")

        Thread {
            try {
                // For now, sync is a simple push + pull
                val pushResult = performPush(authToken)
                val pullResult = performPull(authToken)

                activity.runOnUiThread {
                    call.resolve(JSObject().apply {
                        put("success", true)
                        put("pushedRecords", pushResult)
                        put("pulledRecords", pullResult)
                        put("conflicts", 0)
                        put("timestamp", System.currentTimeMillis())
                    })
                }
            } catch (e: Exception) {
                activity.runOnUiThread {
                    call.reject("Sync failed: ${e.message}")
                }
            }
        }.start()
    }

    // ==================== Push ====================

    @PluginMethod
    fun push(call: PluginCall) {
        val authToken = call.getString("authToken")

        Thread {
            try {
                val pushedRecords = performPush(authToken)

                activity.runOnUiThread {
                    call.resolve(JSObject().apply {
                        put("success", true)
                        put("pushedRecords", pushedRecords)
                    })
                }
            } catch (e: Exception) {
                activity.runOnUiThread {
                    call.reject("Push failed: ${e.message}")
                }
            }
        }.start()
    }

    // ==================== Pull ====================

    @PluginMethod
    fun pull(call: PluginCall) {
        val authToken = call.getString("authToken")

        Thread {
            try {
                val pulledRecords = performPull(authToken)

                activity.runOnUiThread {
                    call.resolve(JSObject().apply {
                        put("success", true)
                        put("pulledRecords", pulledRecords)
                    })
                }
            } catch (e: Exception) {
                activity.runOnUiThread {
                    call.reject("Pull failed: ${e.message}")
                }
            }
        }.start()
    }

    // ==================== Sync Vault ====================

    @PluginMethod
    fun syncVault(call: PluginCall) {
        val userId = call.getString("userId")
        val authToken = call.getString("authToken")

        if (userId == null) {
            call.reject("Missing required parameter: userId")
            return
        }

        Thread {
            try {
                val json = JSONObject().apply {
                    put("userId", userId)
                }

                val requestBody = json.toString().toRequestBody("application/json".toMediaType())
                val requestBuilder = Request.Builder()
                    .url("$backendUrl/api/sync/vault")
                    .post(requestBody)

                if (authToken != null) {
                    requestBuilder.addHeader("Authorization", "Bearer $authToken")
                }

                val response = httpClient.newCall(requestBuilder.build()).execute()
                val success = response.isSuccessful

                activity.runOnUiThread {
                    call.resolve(JSObject().put("success", success))
                }
            } catch (e: Exception) {
                activity.runOnUiThread {
                    call.reject("Vault sync failed: ${e.message}")
                }
            }
        }.start()
    }

    // ==================== Get Sync Status ====================

    @PluginMethod
    fun getSyncStatus(call: PluginCall) {
        // Return current sync status
        // For now, return defaults since we don't have local SQLCipher yet
        call.resolve(JSObject().apply {
            put("pendingCount", 0)
            put("lastSyncTimestamp", 0)
            put("hasPendingChanges", false)
        })
    }

    // ==================== Private Helpers ====================

    private fun performPush(authToken: String?): Int {
        // Placeholder for actual push logic with SQLCipher
        Log.d(TAG, "🔄 [HushhSync] Push completed")
        return 0
    }

    private fun performPull(authToken: String?): Int {
        // Placeholder for actual pull logic with SQLCipher
        Log.d(TAG, "🔄 [HushhSync] Pull completed")
        return 0
    }
}
